File: 0002-add-two-numbers/0002-add-two-numbers.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode AddTwoNumbers(ListNode l1, ListNode l2) {
        ListNode node = l1;
        ListNode node2 = l2;
        
        ListNode sol = new ListNode(0, null);
        ListNode head = sol;
        
        while (true)
        {
            int a = (node != null) ? node.val : 0;
            int b = (node2 != null) ? node2.val : 0;
            int sum = a + b + sol.val;                  
            sol.val = sum % 10;
            int carry = sum / 10;   
            
            if ((node != null && node.next != null) ||
                (node2 != null && node2.next != null) ||
                carry != 0)
            {
                node = node != null ? node.next : null;
                node2 = node2 != null ? node2.next : null;  
                sol.next = new ListNode(carry, null);
                sol = sol.next; 
            }
            else
                break;
        }
        
        return head;
    }
}

File: 0002-add-two-numbers/0002-add-two-numbers.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    var result *ListNode = &ListNode{}
    var cur *ListNode = result
    
    for {
        if (l1 == nil) && (l2 == nil) {
            break
        }
        
        if l1 != nil {
            cur.Val += l1.Val
            l1 = l1.Next
        }
        
        if l2 != nil {
            cur.Val += l2.Val
            l2 = l2.Next
        }
        
        if cur.Val >= 10 {
            var carry int = cur.Val / 10
            cur.Val = cur.Val % 10
            cur.Next = &ListNode{Val: carry}
        } else if (l1 != nil) || (l2 != nil)  {
            cur.Next = &ListNode{}
        }
        cur = cur.Next
    }
    
    return result
}